pkg/settings/bucket.go

```diff
@@ -0,0 +1,148 @@
+	"bytes"
+	"context"
+	"encoding/json"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/thanos-io/objstore"
+	"golang.org/x/exp/slices"
+
+	settingsv1 "github.com/grafana/pyroscope/api/gen/proto/go/settings/v1"
+)
+
+var (
+	oldSettingErr    = errors.New("newer update already written")
+	settingsFilename = "tenant_settings.json"
+)
+
+// NewMemoryStore will create a settings store with an in-memory objstore
+// bucket.
+func NewMemoryStore() (Store, error) {
+	return NewBucketStore(objstore.NewInMemBucket())
+}
+
+// NewBucketStore will create a settings store with an objstore bucket.
+func NewBucketStore(bucket objstore.Bucket) (Store, error) {
+	store := &bucketStore{
+		store:  make(map[string]map[string]*settingsv1.Setting),
+		bucket: bucket,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	store.rw.Lock()
+	defer store.rw.Unlock()
+
+	err := store.unsafeLoad(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return store, nil
+}
+
+type bucketStore struct {
+	rw sync.RWMutex
+
+	// store is kv pairs, indexed first by tenant id.
+	store map[string]map[string]*settingsv1.Setting
+
+	// bucket is an object store bucket.
+	bucket objstore.Bucket
+}
+
+func (s *bucketStore) Get(ctx context.Context, tenantID string) ([]*settingsv1.Setting, error) {
+	s.rw.RLock()
+	defer s.rw.RUnlock()
+
+	tenantSettings := s.store[tenantID]
+
+	settings := make([]*settingsv1.Setting, 0, len(s.store[tenantID]))
+	for _, setting := range tenantSettings {
+		settings = append(settings, setting)
+	}
+
+	slices.SortFunc(settings, func(a, b *settingsv1.Setting) int {
+		return strings.Compare(a.Name, b.Name)
+	})
+	return settings, nil
+}
+
+func (s *bucketStore) Set(ctx context.Context, tenantID string, setting *settingsv1.Setting) (*settingsv1.Setting, error) {
+	s.rw.Lock()
+	defer s.rw.Unlock()
+
+	_, ok := s.store[tenantID]
+	if !ok {
+		s.store[tenantID] = make(map[string]*settingsv1.Setting, 1)
+	}
+
+	oldSetting, ok := s.store[tenantID][setting.Name]
+	if ok && oldSetting.ModifiedAt > setting.ModifiedAt {
+		return nil, errors.Wrapf(oldSettingErr, "failed to update %s", setting.Name)
+	}
+	s.store[tenantID][setting.Name] = setting
+
+	err := s.unsafeFlush(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return setting, nil
+}
+
+func (s *bucketStore) Flush(ctx context.Context) error {
+	s.rw.Lock()
+	defer s.rw.Unlock()
+
+	return s.unsafeFlush(ctx)
+}
+
+func (s *bucketStore) Close() error {
+	return s.bucket.Close()
+}
+
+// unsafeFlush will flush the store to disk. This is not thread-safe, the
+// store's write mutex should be acquired first.
+func (s *bucketStore) unsafeFlush(ctx context.Context) error {
+	data, err := json.Marshal(s.store)
+	if err != nil {
+		return err
+	}
+
+	err = s.bucket.Upload(ctx, settingsFilename, bytes.NewReader(data))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// unsafeLoad will sync the store with an on-disk store, if found. This is not
```

Comment:
```suggestion
// unsafeLoad will sync the store with an object storage store, if found. This is not
```

Reply:
Fixed in https://github.com/grafana/pyroscope/pull/2905/commits/199ad673e5d4516948982adb2a9a20752d69f91d